File: solutions/ruby/resistor-color-trio/2/resistor_color_trio.rb
class ResistorColorTrio
  PREFIX = "Resistor value: "
  SUFFIX = "ohms"
  KILO   = "kilo"
  CODES  = %w[black brown red orange yellow green blue violet grey white].freeze

  attr_reader :colors

  def initialize(colors)
    @colors = colors
  end

  def label
    raise ArgumentError unless valid?
    PREFIX + value + SUFFIX
  end

  private

  def valid?
    colors.size == 3 && colors.all? { |c| CODES.include?(c) }
  end

  def value
    amount = numeric_value
    amount >= 1000 ? text(amount / 1000) + KILO : text(amount)
  end

  def numeric_value
    (digit(0) * 10 + digit(1)) * 10**digit(2)
  end

  def digit(index)
    CODES.index(colors[index])
  end

  def text(amount)
    amount.to_s + " "
  end
end

File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
const BILLION = 1_000_000_000
const MILLION = 1_000_000
const THOUSAND = 1_000

const COLORS: string[] = [
  'black',
  'brown',
  'red',
  'orange',
  'yellow',
  'green',
  'blue',
  'violet',
  'grey',
  'white'
]

interface Simplification {
  amount: number,
  prefix: string
}

const colorValue = (color: string): number => COLORS.indexOf(color)

const baseValue = (colors: string[]): number =>
  10 * colorValue(colors[0]) + colorValue(colors[1])

const multiple = (colors: string[]): number =>
  Math.pow(10, colorValue(colors[2]))

export function decodedResistorValue(colors: string[]): string {
  let simplified = simplify(baseValue(colors) * multiple(colors))
  return `${simplified.amount} ${simplified.prefix}ohms`
}

function simplify(fullAmount: number): Simplification {
  let amount = fullAmount
  let prefix = ''

  if (amount === 0) {
    return { amount, prefix }
  }

  if (fullAmount % BILLION === 0) {
    amount /= BILLION
    prefix = 'giga'
  }
  else if (fullAmount % MILLION === 0) {
    amount /= MILLION
    prefix = 'mega'
  }
  else if (fullAmount % THOUSAND === 0) {
    amount /= THOUSAND
    prefix = 'kilo'
  }

  return { amount, prefix }
}
